### PR TITLE
Fake parts for filling vehicle gaps

### PR DESCRIPTION
**Pull request by kevingranade**
#### Summary
Features "Fill gaps in vehicle walls when necessary to maintain vehicle integrity."

#### Purpose of change
Completes #53722, which is a continuation of #42462, which is an extension of #41277
Fixes https://github.com/CleverRaven/Cataclysm-DDA/issues/5684

As vehicles turn, they skew and open gaps in previously aligned lines of obstacles. This creates a weird situation where a vehicle that is driving in a cardinal direction fully protects its occupants, but a vehicle driving at any other angle can get hitchhikers.

#### Describe the solution
On vehicle refresh, the outline of the vehicle is scanned for eligible obstacles and protuberances that have skewed away from their neighbors, when found these obstacles have fake parts added to the vehicle to fill the gaps.
Vehicle damage code forwards all damage instances to the real part instead of the fake part until the real part is destroyed, at which point the fake part is marked for removal as well.  Updates to the state of the real part are echoed to the fake parts on vehicle refresh as the fake parts are stripped off of the vehicle and recreated.

#### Describe alternatives you've considered
Too many to count, but this approach was selected because it mostly contains the change to the vehicle code instead of sprawling across all of the codebase that has to interact with the map.

#### Testing
Multiple layers of tests were added to exercise this feature, and the propensity for the overmap_terrain_coverage test to trigger several problematic use cases in mapgen has also been extremely valuable.
- [x] have one lingering issue, which is that ramming other vehicles generates `src/vehicle.cpp:7390 [int vehicle::get_non_fake_part(int)] Returning -1 for get_non_fake_part on part_num -1.` errors.
Turns out there was just a typo in the vehicle->vehicle collision code.

#### Additional context
When merging definitely do a merge even with the relatively messy history to preserve authorship.

### CODE REVIEW

**Comment by Inglonias** [comment]
That's a lot of changes. Best of luck

**Comment by kevingranade** [comment]
Ok two major changes:
1. There was a pre-existing issue that came up in testing because I was hammering on it, which is that when a vehicle is placed and then split in mapgen, the split code would call back into the global map instance to place the pieces instead of placing them on the tinymap being generated, the main things causing this were the wreckage generation code, which can cause vehicle splits now, and some map extras which would call map::destroy() and end up splitting vehicles that way.  I fixed this by passing references to the current map being acted on into the vehicle damage functions so that they can target the right map.
2. There was a major problem with the fake parts code in that it would frequently strip the fake parts back out and then replace them, but this really can not happen because doing so invalidates part indexes and causes the state of the vehicle to go out of sync with the map cache. I addressed this by removing calls that would strip fake parts out with the exception of part cleanup code and unracking code, both of which are careful to invalidate and regenerate the vehicle cache afterwards.

I've hammered on the automated tests a ton (50 odd test suite invocations), and I'm pretty confident the errors I was seeing cropping up there periodically are resolved, but I haven't had a chance to do more than some light driving around and ramming stuff testing in-game for some more random testing.

I'll be pushing fixes tot he things clang-tidy is complaining about as soon as I can, but this PR really does need some fairly heavy testing before it goes in.

**Comment by esotericist** [comment]
loaded up my actual save to do for-reals playtesting.

first thing i noticed:
image

my first surprise was that the fake parts are rendering for this flatbed truck when they didn't for the cube van above, but something more pressingly weird is happening with the door fake part:
image

i can only open/close the real door, not the fake one, which... could actually be problematic in some configurations.

back to rendering:
the first time i started up my rv this run, i got correctly rendering proxy parts when i started to turn:
image

then as soon as i turn off the engine:
image

whatever's happening here, it seems to work correctly the first time a vehicle is loaded in a game session, or the first time a vehicle rotates after a game session starts, then breaks once the player is no longer driving the vehicle until a new session. (turning off the engine is not necessary; i was able to reproduce the 'no tile' result above by simply letting go of the controls while the vehicle was otherwise stopped).

so far i have yet to discern any way of getting it back to the valid state other than save and quit, then restart.

File: src/vehicle.h [hunk 1]
```diff
@@ -1858,13 +1896,25 @@ class vehicle
...
+        // Removes fake parts from the parts vector
+        void remove_fake_parts( bool cleanup = true );
+        tripoint get_abs_diff( const tripoint &one, const tripoint &two ) const;
+        bool should_enable_fake( const tripoint &fake_precalc, const tripoint &parent_precalc,
+                                 const tripoint &neighbor_precalc ) const;
+        /**
+        *  checks carried_vehicles param for duplicate entries of bike racks/vehicle parts
         * this eliminates edge cases caused by overlapping bike_rack lanes
-        * @param carried_vehicles is a set of either vehicle_parts or bike_racks that need duplicate entries across the vector<vector>s rows removed
+        * @param carried_vehicles is a set of either vehicle_parts or bike_racks that need duplicate entries accross the vector<vector>s rows removed
```
Comment (BrettDong) [inline src/vehicle.h:1914]:
```suggestion
        * @param carried_vehicles is a set of either vehicle_parts or bike_racks that need duplicate entries across the vector<vector>s rows removed
```

**Comment by kevingranade** [comment]
I need to flip to a tiles build to double check for the specific issues you're reporting, but I did find a critical error that is almost certainly related.
There's a cache of vehicle-relative coordinates to item index called relative_parts that is updated in vehicle::refresh(), when I overhauled fake part maintenance to avoid removing and re-adding fake parts on refresh (to prevent map cache desync), I failed to have refresh() rebuild relative_parts, so it would lose track of the fake parts on some kinds of refresh operation and then re-add them on other flavors of refresh.
I also need to clean up the code, what I have is working, but it's very messy.

**Comment by kevingranade** [comment]
I've finally had a chance to do a tiles build and test, and I'm working my way through the issues you reported.
1. A diagonally oriented cube van's mirrors are rendered properly, This seems to be related to the relative_parts cache and is fixed.
2. "the rendering needs a band-aid, and x look around shows that there's a vehicle there but not what the tile is supposed to be, which is awkward."  This issue is still present, I'm pretty sure it needs a separate override in look_around.
3. Fake parts are transparent, or rather are not participating in transparency.
4. Inability to access doors via their fake parts is still occurring.

**Comment by andrei8l** [comment]
<details>
<summary>Collisions sometimes (temporarily) remove real parts</summary>

Screenshot from 2022-05-03 22-03-26
</details>

<details>
<summary>Another try from same angle</summary>

Screenshot from 2022-05-03 22-05-27
</details>

<details>
<summary>And it's not just a visual bug</summary>

Screenshot from 2022-05-03 22-04-48
</details>

<details>
<summary>But the missing parts do come back after moving a bit</summary>

Screenshot from 2022-05-03 22-05-36
</details>

[Forgan.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/8612966/Forgan.tar.gz)

**Comment by esotericist** [comment]
> Collisions sometimes (temporarily) remove real parts

that looks more or less like the standard vehicle cache consequences of getting two vehicles to overlap. are fake parts maybe getting placed without checking for collision before they were placed? i would imagine the rotation code would have trouble checking for collisions on fake parts sometimes when they wouldn't be present until after the rotation evaluates

**Comment by kevingranade** [comment]
> are fake parts maybe getting placed without checking for collision before they were placed?

That sounds likely, all kinds of bad things happen when two vehicles try to occupy the same map location. I'll see if I can reproduce in a controlled way.

**Comment by kevingranade** [comment]
Yea that looks like cache overwriting.

**Comment by kevingranade** [comment]
It is cache overwriting, but the collision is a red herring, turns out what's happening is when refreshing fake parts, we weren't careful enough to make sure they were "active" fakes, so we were clearing vehicle caches (which wouldn't be refreshed for a while) even on fakes that are not currently being rendered.

**Comment by kevingranade** [comment]
Ok last call for testing, I've adressed all the known issues and I'm not going to let this stall out at the end just because it's too scary.

I'll be merging it tomorrow unless it gets more testing feedback.

**Comment by andrei8l** [comment]
<details>
<summary>You can close doubled doors if only one part is blocked</summary>

https://user-images.githubusercontent.com/68240139/169733623-21ed32e8-51f4-44bc-9a42-6f9090b99abf.mp4
</details>

**Comment by kevingranade** [comment]
> door bug.

Thanks! Reproduced locally, should be a simple fix.

File: data/mods/Dark-Skies-Above/mapgen/map_extras/patrol_test.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
+[
+  {
+    "type": "mapgen",
+    "method": "json",
+    "update_mapgen_id": "mx_dsa_patrol_test",
```
Comment (Night-Pryanik) [inline data/mods/Dark-Skies-Above/mapgen/map_extras/patrol_test.json:5]: This seems unrelated to fake vehicle parts stuff?

File: src/mapgen.cpp [hunk 3]
```diff
@@ -6517,45 +6519,63 @@ std::unique_ptr<vehicle> map::add_vehicle_to_map(
...
                 std::vector<vehicle_part *> parts_to_move = veh_to_add->get_parts_at( map_pos, "",
                         part_status_flag::any );
                 if( !parts_to_move.empty() ) {
                     // Store target_point by value because first_veh->parts may reallocate
                     // to a different address after install_part()
-                    const point target_point = first_veh->get_parts_at( map_pos, "",
-                                               part_status_flag:: any ).front()->mount;
+                    std::vector<vehicle_part *> first_veh_parts = first_veh->get_parts_at( map_pos, "",
+                            part_status_flag:: any );
+                    // this happens of this location is occupied by a fake part.
```
Comment (Night-Pryanik) [inline src/mapgen.cpp:6532]:
```suggestion
                    // this happens if this location is occupied by a fake part.
```

**Comment by Night-Pryanik** [comment]
When installing vehicle parts over fake parts, real parts get fake part for free:

Disposition: highlighted part (with a headlight sprite) is a fake part with fake quarterpanel
изображение

Nothing is here in reality
изображение

Installing a blade here
изображение

Blade is installed here
изображение

But also here
изображение

Straightening the car reveals that fake quarterpanel part manifested in reality
изображение

**Comment by Night-Pryanik** [comment]
You can still close real doors when standing on fake doors, and vice versa.

File: data/mods/Dark-Skies-Above/mapgen/map_extras/patrol_test.json (hunk 2 above)
Comment (kevingranade) [reply data/mods/Dark-Skies-Above/mapgen/map_extras/patrol_test.json:5]: Thanks for reminding me, this might be a merge error or something, I need to look into reverting it.

**Comment by kevingranade** [comment]
Door issue fixed and backstopped by a test.
I still need to take a look at the odd mod adjustments and the installed part issue.

**Comment by kevingranade** [comment]
Installation issue, which is actually a display issue is resolved.
look_around was showing inactive fake parts.
The vehicle interact menu was showing fake parts as long as they're on the same tile as a real part.
